doc/Language/iterating.pod6

```diff
@@ -4,136 +4,321 @@
-        self.bless( :$chain );
+    method new(Str $chain where { .match(/^^ <[ACGT]>+ $$/) && .chars %% 3 }) {
+        self.bless(:$chain);
     }
 
-    method iterator( ){ self }
-    method pull-one( --> Mu){
-        if $!index < $.chain.chars {
-            my $codon = $.chain.comb.rotor(3)[$!index div 3];
-            $!index += 3;
-            return $codon;
-        } else {
-            return IterationEnd;
+    method iterator(DNA:D:) { self }
+    method pull-one(--> Mu) {
+        return IterationEnd unless 3 * $!index < $!chain.chars;
+
+        my $codon = do given $!chain.comb.rotor(3)[$!index] {
+            when $!index %% 2 { $_.reverse }
+            when $!index %% 3 { $_.tail, |$_.head(2) }
+            default           { $_ }
         }
+
+        $!index += 1;
+        return $codon.list;
     }
-};
+}
+=end code
+
+There are two main differences between this implementation and the
+implementation from the previous section:
+
+=item The C<iterator> method returns a single instance of the class per instance
+object. In the previous section, the C<iterator> method returns a brand new
+C<DNAIterator> object on demand per instance object. 
+
+=item The class itself is an iterator. Unlike iterables that are I<stateless>,
+iterators are I<stateful> (i.e., once an item has been consumed from them, it's
+gone), an instance object of the class will return the same iterator (i.e., the
+class itself) which won't produce more values once they've been consumed. One
+way to mitigate this is to let an array consume the instance object's values,
+and then use that array as needed. Another alternative is to reset C<$!index> to
+0 once all the values are consumed, however this would violate the iterator
+protocol.
 
-my $a := DNA.new('GAATCC');
-.say for $a; # OUTPUT: «(G A A)␤(T C C)␤»
+=begin code :skip-test<compile time error; undeclared name>
+my \dna = DNA.new('ACGTACGTT');
+
+say dna.map(*.join).join('|'); # OUTPUT: «GCA|TAC|TTG␤»
+say dna.map(*.join).join('|'); # OUTPUT: «␤»
 =end code
 
-We declare a C<DNA> class which does the two roles, C<Iterator> and C<Iterable>;
-the class will include a string that will be constrained to have a length that
-is a multiple of 3 and composed only of ACGT.
-
-Let us look at the
-C<pull-one> method. This one is going to be called every time a new iteration
-occurs, so it must keep the state of the last one. An C<$.index> attribute will
-hold that state across invocations; C<pull-one> will check if the end of the
-chain has been reached and will return the C<IterationEnd> constant provided by
-the role. Implementing this low-level interface, in fact, simplifies the
-implementation of the C<Iterable> interface. Now the iterator will be the object
-itself, since we can call C<pull-one> on it to access every member in turn;
-C<.iterator> will thus return just C<self>; this is possible since the object
-will be, at the same time, C<Iterable> and C<Iterator>.
-
```

Comment:
Don't think it's a good idea to just delete content where there's no issue. If there's an error here, point it out. If there's no error, try to repurpose it or extend it so that it fits better your narrative.